api/client/client.go

```diff
@@ -272,21 +272,28 @@ func connect(ctx context.Context, cfg Config) (*Client, error) {
 				})
 			}
 
-			// Connect with dialer provided in creds.
-			if dialer, err := creds.Dialer(cfg); err != nil {
-				if !trace.IsNotImplemented(err) {
-					sendError(trace.Wrap(err, "failed to retrieve dialer from creds of type %T", creds))
-				}
-			} else {
-				syncConnect(ctx, dialerConnect, connectParams{
-					cfg:       cfg,
-					tlsConfig: tlsConfig,
-					dialer:    dialer,
+			addrs := cfg.Addrs
+			if len(addrs) == 0 {
+				// If there's no explicitly specified address, fall back to
+				// an address provided by the credential, if it provides one.
+				credAddrSource, ok := creds.(interface {
+					ProxyWebAddr() (string, error)
 				})
+				if ok {
+					addr, err := credAddrSource.ProxyWebAddr()
+					if err != nil {
+						sendError(trace.Wrap(err))
+						continue
+					}
+					addrs = []string{addr}
```

Comment:
> A minor downside to this approach is that we are now trying to connect to what we know is a proxy address as an

I'm actually not entirely convinced that this is a downside. When I tried connecting using just the Profile Credential dialer to my cluster, which is a fairly out of the box TLS routing enabled cluster, this failed. The type of dialer that the credential uses requires that the cluster has a separate or multiplexed SSH port (not just ALPN routed). Since switching to my implementation, the address and credentials sourced from the Profile can be used with a different technique and can connect succesfully.

Reply:
> When I tried connecting using just the Profile Credential dialer to my cluster, which is a fairly out of the box TLS routing enabled cluster, this failed. The type of dialer that the credential uses requires that the cluster has a separate or multiplexed SSH port (not just ALPN routed). Given the fairly common/standard setup that my cluster uses, I'm lead to believe that using the Profile credential with no explicitly specified address probably doesn't work for most customers anyway.

Well that sounds like a bug :smile: the profile credentials definitely should have worked out of the box, but I agree trying all methods is fine and potentially avoids pigeon-holing us.